<center>
<h1> TP-Projet d'optimisation numérique </h1>
<h1> Algorithme du lagrangien augmenté </h1>
</center>

## Implémentation

1. Implémenter l'algorithme du lagrangien augmenté, en utilisant les différentes méthodes
qui ont été vues en première partie pour la résolution de la suite de problèmes sans
contraintes (fichier `src/lagrangien_augmente.jl`). La spécification est donnée dans le fichier.
2. Vérifier que les tests ci-dessous passent.
 

In [7]:
include("../src/lagrangien_augmente.jl")
include("../test/tester_lagrangien_augmente.jl")

#
afficher = true# si true, alors affiche les résultats des algorithmes

#
tester_lagrangien_augmente(lagrangien_augmente, afficher);

Affichage des résultats des algorithmes : true

Coefs : 4.499961817889869 80
-------------------------------------------------------------------------
Résultats de : LA et newton appliqué à f1:
  * x0       = [0, 1, 1]
  * x_sol    = [0.5000042424566808, 1.2499978787716595, 0.5000042424566811]
  * f(x_sol) = 2.249961818051857
  * nb_iters = 7
  * flag     = 0
  * solution = [0.5, 1.25, 0.5]
Coefs : 4.499961817889869 80
-------------------------------------------------------------------------
Résultats de : LA et newton appliqué à f1:
  * x0       = [0.5, 1.25, 1.0]
  * x_sol    = [0.5000042424566808, 1.2499978787716595, 0.5000042424566811]
  * f(x_sol) = 2.249961818051857
  * nb_iters = 7
  * flag     = 0
  * solution = [0.5, 1.25, 0.5]
Coefs : 0.03865150236679149 40
-------------------------------------------------------------------------
Résultats de : LA et newton appliqué à f2:
  * x0       = [1, 0]
  * x_sol    = [0.9072336571162266, 0.8227549044924193]
  * f(x_sol) = 0.0086157070

## Interprétation

 1. Commenter les résultats obtenus, en étudiant notamment les valeurs en sortie de $\lambda_k$ et $\mu_k$.
 2. Étudier l'influence du paramètre $\tau$ dans la performance de l'algorithme. Pour cela vous réaliserez des tests numériques.

In [ ]:
# Expérimentations numériques à faire ici.
# Vous pouvez utiliser le package Plots pour les affichages de courbes: using Plots
include("../test/fonctions_de_tests.jl")
include("../src/lagrangien_augmente.jl")

val_τ = range(2,0.5,3)
t_τ = []

# Série de tests n°1 : avec le cas de test qui prend le plus d'itérations
x0 = pts2.x01
tol = 1e-5

for τ in val_τ
    t = time()
    x_sol, f_sol, flag, nb_iters, _ = lagrangien_augmente(fct1, grad_fct1, hess_fct1, contrainte1, grad_contrainte1, hess_contrainte1, x0, tol_abs=tol, tol_rel=tol, algo_noc="rc-cauchy",τ = τ)
    if flag == 3
        erreur = true
    end
    t_τ = vcat(t_τ, [time()-t])
end
print(t_τ)



Coefs : 4.499999289228105

### Réponses
 1. Pour une fonction donnée, $\lambda_k$ (final) va être le même; tandis que pour $\mu_k$, il peut avoir beaucoup augmenté avec le nombre d'itérations pour ramener l'optimisation dans la zone de contrainte ($\mu_k$ sert à pénaliser $||c||$).
 En particulier pour le cas de test "LA et rc-cauchy appliqué à f1" avec x0 = [0.5, 1.25, 1.0], on a 52 itérations et $\mu_k = 1310720$
 2. $\tau$ intervient dans la modification du facteur de pénalité $\mu_k$ dans le cas où on ne se trouve pas dans la zone de contrainte.